### PR TITLE
Improving PowerLaw implementation

### PR DESCRIPTION
**Pull request by InfinityMod**
**Introduction**
Hi, I also need the PowerLaw for my research in numpyro. I investigated the problem over the last two days as to why the current implementation of the power law fails to pass the tests. The formulas are already correctly implemented.

**The problem**
To my mathematical understanding, the different gradients at $\alpha=-1$  stem from a discontinuity point on the transition from the outer bounds ($\alpha=-1+\delta t$ with $\lim_{t \to 0}$) to ($\alpha=-1$) for the normalized probability as for $\alpha\neq-1$ the partial derivation with respect to $\alpha$ of the base equation: 
 $\frac{(α + 1)x^α}{(b^{α + 1} - a^{α + 1})}$
 tends to another value than of the base equation for $\alpha=-1$ :
$\frac{x^α}{\ln(b) - \ln(a)}$.
For the CDF and the ICDF, this issue is rooted in this cause, too. So, the issue is purely mathematical and does not account for the original code or the framework.

**My solution**
I partially integrated (partly via Wolfram Alpha) the log_prob, cdf, and icdf formula with respect to $x, \alpha, a, b$ and defined an individual underlying VJP rule for each formula. For the derivative with respect to $\alpha$, I approximated the value by calculating the borders of the discontinuity point with $-1\pm\delta t; \delta t = 10e-4$ and averaging the two sides to approximate an adequate middle declination. 

Additionally, some improvements in the tests and the activation of float64 (without it, it was too imprecise to pass the tests) have been added too.

Best,
David

### CODE REVIEW

**Review by Qazalbash — APPROVED** [review]
Thank you for implementing the changes. I'm happy the test cases are passing—great work! Before merging, I believe a few areas in the code could be simplified for better clarity and maintainability. Additionally, if you have any resources or insights on how you approached or understood the implementation of the custom derivatives, I would greatly appreciate it if you could share them. Thanks!

**Comment by Qazalbash** [comment]
@InfinityMod Upon some closer look I see many terms are more simplifiable, or did you leave that for computational purposes?

**Comment by InfinityMod** [comment]
@Qazalbash Simplification would be welcome. I didn't find a further way to simplify. I cross-checked the calculations with Wolfram Alpha and guessed they were already as simple as possible. If there's a way, a simplification would be pretty welcome.

_Wolfram Alpha command for the normalized probability and_ $\alpha \neq 1$:
Jacobian matrix | (x^α)/(log(b) - log(a)) | with respect to | (x, α, a, b)

image

_Wolfram Alpha command for the normalized probability and_ $\alpha =1$:
Jacobian matrix | log(x)/(log(b) - log(a)) | with respect to | (x, a, b)

image

_Wolfram Alpha command for CDF and_ $\alpha \neq 1$:
Jacobian matrix | (x^(α + 1) - a^(α + 1))/(b^(α + 1) - a^(α + 1)) | with respect to | (x, α, a, b)

image

_Wolfram Alpha command for the CDF and_ $\alpha =1$:
Jacobian matrix | (log(x) - log(a))/(log(b) - log(a)) | with respect to | (x, a, b)

image

_Wolfram Alpha command for iCDF and_ $\alpha \neq 1$:
Jacobian matrix | (a^(1 + α) + q (b^(1 + α) - a^(1 + α)))^(1/(1 + α)) | with respect to | (q, α, a, b)

image

_Wolfram Alpha command for the iCDF and_ $\alpha =1$:
Jacobian matrix | a (b/a)^q | with respect to | (q, a, b)

image

The formulas are implemented accordingly and applied via "custom_jvp".
If more documentation is needed, please let me know.

**Comment by Qazalbash** [comment]
@InfinityMod By simplification some terms were being cancelled and the other aspect was memory and computation point of view. I have made changes to them you can see them in ce3c53fce08c7f107eabae7a700e7c3c283106cf. The mentioned commit contains all the changes that are meant to reuse quantities.

**Comment by Qazalbash** [comment]
All test cases that were previously been failing are passing, but some other test cases (not related to power laws) are failing which was not the case before. I suspect this is due to the high floating point precession mode. What do you think?

**Comment by InfinityMod** [comment]
@Qazalbash, I see what you mean; well understood :). Thanks for making it less expensive calculation-wise. I didn't optimize at this stage. 

~~Concerning the tests: could a turn 64-bit on/off flag for the test cases help?
E.g. T(dist.DoublyTruncatedPowerLaw, -1.0, 1.0, 2.0, force_64bit=True)~~
-> jax_enable_x64 is unfortunately just considered at initialization, but changes are not reflected later.
-> We could force an x64 calculation inside the function if possible. This increases the complexity a bit.

**Comment by Qazalbash** [comment]
@InfinityMod I tried this on my local machine but it is not powerful enough to handle all these tests. I was monitoring the [automated tests](https://github.com/pyro-ppl/numpyro/actions/runs/10439361332/job/28907944185?pr=1807) on GitHub Actions and found that the test was stuck on a single case for several hours after which it was cancelled. It would be great if you could look into it. This PR has been merged. Can you reflect your changes on the main [PR](https://github.com/pyro-ppl/numpyro/pull/1807)?

**Comment by Qazalbash** [comment]
> -> jax_enable_x64 is unfortunately just considered at initialization, but changes are not reflected later.
> -> We could force an x64 calculation inside the function if possible. This increases the complexity a bit.

I think you are referring to [`jax.experimental.enable_x64`](https://jax.readthedocs.io/en/latest/_autosummary/jax.experimental.enable_x64.html).

**Comment by InfinityMod** [comment]
Yeah, but it seems to me that this flag can't be switched individually during runtime.
I identified the stuck test; it's test_vmapped_binomial_p0. It seems to have an endless loop on execution with 64bit in "_binomial" at .\distributions\util.py. Ironically it is a test to check it doesn't loop infinite due to the comment:
test that vmapped binomial with p = 0 does not have an infinite loop

**Comment by Qazalbash** [comment]
If you are confident enough to fix it, can you share the fixed implementation?

**Comment by Qazalbash** [comment]
Hi @InfinityMod,

I hope you're doing well. I was reviewing the custom derivatives in this PR and noticed something in [this comment](https://github.com/Qazalbash/numpyro/pull/1#issuecomment-2295341573), and I have found the following mistakes,

- It seems the normalization constant used for the \$\alpha \neq -1\$ case corresponds to the \$\alpha = -1\$ case.
- Normalized pdf for $\alpha\neq-1$ case is incorrect. It should be $\left(x(\ln(b)-\ln(a))\right)^{-1}$
- The Jacobians appear to be for the PDF rather than its logarithm, which the method is meant to implement.

Please let me know if I’ve misunderstood anything.
